src/NRules/NRules.Fluent/Dsl/QueryExtensions.cs

```diff
@@ -108,6 +108,34 @@ public static IQuery<TSource> Where<TSource>(this IQuery<TSource> source, params
             return new QueryExpression<IGrouping<TKey, TElement>>(source.Builder);
         }
 
+        /// <summary>
+        /// Aggregates matching facts into a collection sorted ascending by key.
+        /// </summary>
+        /// <typeparam name="TSource">Type of source facts.</typeparam>
+        /// <typeparam name="TKey">Type of sorting key.</typeparam>
+        /// <param name="source">Query expression builder.</param>
+        /// <param name="keySelector">Key selection expression used for sorting.</param>
+        /// <returns>Query expression builder.</returns>
+        public static IQuery<IEnumerable<TSource>> OrderBy<TSource, TKey>(this IQuery<TSource> source, Expression<Func<TSource, TKey>> keySelector)
```

Comment:
Definitely agree! It's weird ordering without first collecting, and it's nice to get started on this pattern for if stuff like ToDictionary are added.

What if they want to Flatten etc. and then sort? Would that be done via .SelectMany().Collect(),OrderBy()?

I had also considered it briefly but wasn't sure how that would fit into the canonical method, would the modifier be entirely handled in the DSL? Would it have a single SortedAggregator node or both a CollectionAggregator and a SortedAggregator? 

While I have an idea of how this could work for .ThenBy (as discussed in the issue), it's not as clear for this -- makes sense to make the "OrderBy" or "OrderByDesc" heavy and handle "ThenBy"/"ThenByDesc", but not so much doing the same for everything that could modify .Collect()

Reply:
Any updates? Did you decide against this or want to do it in a future release and/or commit?

Reply:
No, definitely want to do this. I just wanted to bank what you have, so that I can also merge my feature branch before it deviates too much.
Regarding SelectMany, yes, they can Collect and OrderBy again after it.
As far as implementation of stacking - could be either separate elements at the canonical model level, or just a Collect with additional expressions. I’m kind of leaning towards the later. What do you think?

Reply:
Not 100% sure I understand, by separate elements do you mean aggregate nodes in the Rete graph and for the other approach are you suggesting canonical method could just have additional Collect methods (e.g. one for single key SortedAggregator, one for multi key SortedAggregator, one for ToDictionary etc.)?

My concern with that approach is that we could end up a case where the same arguments are needed for a different purpose, e.g. single key sorted aggregator and ToDictionary. If that is what you meant, then maybe the pragmatic approach is having a single Rete node but giving the method a name like "CollectAndOrderBy" in canonical model?

If we went with the separate node approach, could multiple modifiers end up sharing the collection CollectionAggregator node? I haven't looked enough into how it all works, but even if it could I guess it would be using a bit of memory for not much gain.

Reply:
There are three levels here - the Rete level, canonical model level (aka RuleModel or AST) and the DSL level.  So, let's discuss them one by one to avoid confusion.
1. At the Rete level I think most of these things will be different aggregator implementations, like you already did - one for plain collect, one for sorted collection, one for a dictionary collect, etc.
2. At the canonical model level, they are all represented by an AggregateElement. The question here is whether they should be different types of aggregators (e.g. Collect vs Sort) or just one and the same (say "Collect") but with different expressions in the ExpressionMap. That's what I was talking about in my previous comment - I'm leaning more towards just using "Collect" aggregator here, with different expressions. TBC, they would still compile to different aggregators at the Rete level.
3. At the DSL level we want these new extension methods just modifying a preceding Collect, so Collect will return something like `ICollectQuery<T>`, so that the new extension methods are only available there.

Does this make sense?

Reply:
That makes sense, good to have some clarity on all the terms now -- thank you!

I definitely agree now about having one "Collect" element for aggregators that modify the collect operation, I'll try changing SortedAggregator over to that approach.

Reply:
Marking this as resolved. We can continue the conversation under #171